walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,126 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public static final int DAY_START = TimeRange.START_OF_DAY;
+    public static final int DAY_END = TimeRange.END_OF_DAY;
+    public static final TimeRange WHOLE_DAY = TimeRange.WHOLE_DAY;
+
+    /**
+      * Return a collection of TimeRange that attendees can meet for a requested 
+      * meeting. If there is not enough room for optional attendees, find TimeRanges 
+      * that just work for required attendees. If there are no required attendees, 
+      * find the TimeRangs for just the optional attendees.
+      */
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        long meetingDuration = request.getDuration();
+        // Required attendees for the request meeting.
+        Collection<String> requiredAttendees = request.getAttendees(); 
+        // All attendees (required and optional) for the request meeting.
+        Collection<String> allAttendees = new ArrayList<>(requiredAttendees);
+        allAttendees.addAll(request.getOptionalAttendees());
+
+        // If the meeting duration is invalid, return an empty Collection.
+        if (meetingDuration > WHOLE_DAY.duration() || meetingDuration <= 0) {
+            return Arrays.asList();
+        }
+
+        // Find available TimeRange for both required and optional attendees.
+        Collection<TimeRange> allMeetingTimes = 
+            findMeetingTimes(events, allAttendees, meetingDuration);
+
+        // If there is meeting time for all, return allMeetingTimes so that 
+        // both required and optional can attend the meeting. 
+        if (allMeetingTimes.size() > 0) {
+            return allMeetingTimes;
+        }
+
+        // If there is no required attendee, return allMeetingTimes for 
+        // optional attendees.
+        if (requiredAttendees.isEmpty()) {
+            return allMeetingTimes;
+        }
+
+        // If not, return available TimeRange for required attendees.
+        return findMeetingTimes(
+            events, requiredAttendees, meetingDuration);
+    }
+
+    /**
+      * Find all times that the meeting could happen, given attendees and duration. 
+      * We loop through all given events to find all available TimeRanges. Pointer 
+      * rangeStart is pointing at the start of the current available TimeRange, which 
+      * is the time between the pointer and start time of the current event.If the current 
+      * available TimeRange has longer duration than the requested meeting duration, add
+      * it to available meetingTimes. Then rangeStart is updated to the end of the current
+      * event.
```

Comment:
Can you try to rephrase this to make it easier to read?

Reply:
I rephrase the whole paragraph and hope it sounds a bit more clear to you! I'll just leave this conversation unresolved in case it needs more clarifications.